app/org/sagebionetworks/bridge/models/ClientInfo.java

```diff
@@ -0,0 +1,272 @@
+        result = prime * result + Objects.hashCode(osVersion);
+        result = prime * result + Objects.hashCode(sdkName);
+        result = prime * result + Objects.hashCode(sdkVersion);
+        return result;
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+        if (this == obj)
+            return true;
+        if (obj == null || getClass() != obj.getClass())
+            return false;
+        ClientInfo other = (ClientInfo) obj;
+        return Objects.equals(appName, other.appName) && Objects.equals(appVersion, other.appVersion)
+                && Objects.equals(osName, other.osName) && Objects.equals(osVersion, other.osVersion)
+                && Objects.equals(sdkName, other.sdkName) && Objects.equals(sdkVersion, other.sdkVersion);
+    }
+
+    @Override
+    public String toString() {
+        return "ClientInfo [appName=" + appName + ", appVersion=" + appVersion + ", osName=" + osName + ", osVersion="
+                + osVersion + ", sdkName=" + sdkName + ", sdkVersion=" + sdkVersion + "]";
+    }
+
+    public static class Builder {
+        private String appName;
+        private Integer appVersion;
+        private String osName;
+        private String osVersion;
+        private String sdkName;
+        private Integer sdkVersion;
+
+        public Builder withAppName(String appName) {
+            this.appName = appName;
+            return this;
+        }
+        public Builder withAppVersion(Integer appVersion) {
+            this.appVersion = appVersion;
+            return this;
+        }
+        public Builder withOsName(String osName) {
+            this.osName = osName;
+            return this;
+        }
+        public Builder withOsVersion(String osVersion) {
+            this.osVersion = osVersion;
+            return this;
+        }
+        public Builder withSdkName(String sdkName) {
+            this.sdkName = sdkName;
+            return this;
+        }
+        public Builder withSdkVersion(Integer sdkVersion) {
+            this.sdkVersion = sdkVersion;
+            return this;
+        }
+        /**
+         * It's valid to have a client info object with no fields, if the
+         * User-Agent header is not in our prescribed format.
+         */
+        public ClientInfo build() {
+            return new ClientInfo(appName, appVersion, osName, osVersion, sdkName, sdkVersion);
+        }
+
+    }
+    
+    /**
+     * Get a ClientInfo object given a User-Agent header string. These values are cached and 
+     * headers that are not in the prescribed format return an empty client info object.
+     * @param userAgent
+     * @return
+     */
+    public static ClientInfo fromUserAgentCache(String userAgent) {
+        if (!StringUtils.isBlank(userAgent)) {
+            try {
+                return userAgents.get(userAgent);    
+            } catch(ExecutionException e) {
+                logger.debug(e.getMessage(), e);
```

Comment:
Reading through the code, it looks like this exception could never happen. Worth a comment here saying this could never happen?

If it does happen, should we log an error instead of debug so we can figure out why?